src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt

```diff
@@ -0,0 +1,206 @@
+    )
+    private val desiredLevelPatter by patternGroup.pattern(
+        "slot24.name.level",
+        "Desired Level: (200|1?[0-9]?[0-9])",
+    )
+    private val userInputPattern by patternGroup.pattern(
+        "slot24.name.input",
+        "User Input",
+    )
+    private val trainingTypePattern by patternGroup.pattern(
+        "training.type",
+        "Type: (Free|Light|Moderate|Expert|Ultra|Turbo!)",
+    )
+
+    @SubscribeEvent
+    fun onFannAnvilTooltip(event: LorenzToolTipEvent) {
+        if (!trainingSlotInventoryPattern.matches(InventoryUtils.openInventoryName())) return
+        if (!anvilPattern.matches(event.itemStack.displayName)) return
+        val tooltip = event.toolTip
+
+        val trainingType = tooltip.getTrainingType() ?: return
+
+        when (trainingMode) {
+            TrainingMode.DAY_COUNT -> {
+                if (trainingType == TrainingType.FREE) return
+                if (!showCoins || !showBits) return
+
+                val totalExp = tooltip.getExpEarned() ?: return
+                val coinPerExp = tooltip.getCoins() / totalExp
+                val xpPerBit = totalExp / tooltip.getBits()
+
+                tooltip.insertLineAfter(coinsPattern, "   §6Coins/XP: ${coinPerExp.roundTo(2)}")
+                tooltip.insertLineAfter(bitsPattern, "   §6XP/Bit: ${xpPerBit.roundTo(2)}")
+
+            }
+
+            TrainingMode.UNTIL_LEVEL -> {
+                if (trainingType == TrainingType.FREE) return
+                if (!showCoins || !showBits) return
+
+                val dailyExp = tooltip.getDailyExp() ?: return
+                val duration = tooltip.getDuration() ?: return
+                val totalExp = dailyExp * duration
+                val coinPerExp = tooltip.getCoins() / totalExp
+                val xpPerBit = totalExp / tooltip.getBits()
+                tooltip.insertLineAfter(coinsPattern, "   §6Coins/XP: ${coinPerExp.roundTo(2)}")
+                tooltip.insertLineAfter(bitsPattern, "   §6XP/Bit: ${xpPerBit.roundTo(2)}")
+
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onInventoryUpdate(event: InventoryUpdatedEvent) {
+        if (!trainingSlotInventoryPattern.matches(event.inventoryName.removeColor())) return
+        val slot24 = event.inventoryItems[24]
+        if (slot24 != null) {
+            val name = slot24.displayName.removeColor()
+            if (desiredLevelPatter.matches(name)) {
+                trainingMode = TrainingMode.UNTIL_LEVEL
+            } else if (userInputPattern.matches(name)) {
+                trainingMode = TrainingMode.DAY_COUNT
+            }
+        }
+    }
+
+    private fun MutableList<String>.insertLineAfter(pattern: Pattern, content: String) {
+        val iter = this.listIterator()
+        while (iter.hasNext()) {
+            val line = iter.next().removeColor()
+            if (pattern.matcher(line).find()) {
+                iter.add(content)
+            }
+        }
+    }
```

Comment:
This seems like a cool function that could be used elsewhere in the mod, maybe move it to a utils class such as collection utils